Convert LinearAlgebra::distributed::Vector to Kokkos

masterleinad opened the pull request:
Depends on #14510.

masterleinad commented:
/rebuild

tjhei commented:
Can you rebase to see if the new MPI Jenkins passes?

Rombur commented on include/deal.II/lac/vector_operations_internal.h:
```diff
@@ -2472,32 +2478,22 @@ namespace internal
+          KOKKOS_LAMBDA(size_type i, Number & update) {
+            update += data.values(i);
+          },
+          result);
```
The name of the function is `mean_value`, shouldn't we divide the `result` by the size?

Rombur commented on include/deal.II/lac/vector_operations_internal.h:
```diff
@@ -2508,44 +2504,55 @@ namespace internal
+#elif KOKKOS_VERSION < 30700
+            update += Kokkos::Experimental::fabs(data.values(i));
+#else
+            update += Kokkos::fabs(data.values(i));
```
```suggestion
            update += Kokkos::abs(data.values(i));
```

Rombur commented on include/deal.II/lac/vector_operations_internal.h:
```diff
@@ -2508,44 +2504,55 @@ namespace internal
+            update += Kokkos::Experimental::pow(
+              Kokkos::Experimental::fabs(data.values(i)), exp);
+#else
+            update += Kokkos::pow(Kokkos::fabs(data.values(i)), exp);
```
```suggestion
            update += Kokkos::pow(Kokkos::abs(data.values(i)), exp);
```

Rombur commented on include/deal.II/lac/vector_operations_internal.h:
```diff
@@ -2625,27 +2627,26 @@ namespace internal
+              Kokkos::subview(data.values,
+                              Kokkos::pair<size_type, size_type>(0, size)),
+              Kokkos::subview(v_data.values,
+                              Kokkos::pair<size_type, size_type>(0, size)));
```
Don't think you need the subview

Rombur commented on source/lac/la_parallel_vector_cuda.cc:
```diff
@@ -24,40 +24,50 @@ namespace LinearAlgebra
-    template class Vector<float, ::dealii::MemorySpace::CUDA>;
-    template class Vector<double, ::dealii::MemorySpace::CUDA>;
+    template class Vector<float, ::dealii::MemorySpace::Default>;
+    template class Vector<double, ::dealii::MemorySpace::Default>;
```
Is this file still necessary? We could do this in `la_parallel_vector.inst.in`

masterleinad commented on include/deal.II/lac/vector_operations_internal.h:
```diff
@@ -2111,103 +2110,98 @@ namespace internal
+              Kokkos::subview(data.values,
+                              Kokkos::pair<size_type, size_type>(0, size)),
+              Kokkos::subview(v_data.values,
+                              Kokkos::pair<size_type, size_type>(0, size)));
```
I don't think this would be correct. We use this function to copy only the locally owned part of the vector.

Rombur commented:
Code doesn't compile anymore. Did you mean to push the new commits from Lauren's account?

masterleinad commented:
> Code doesn't compile anymore. Did you mean to push the new commits from Lauren's account?

I fixed the authorship for all commits and reverted unifying the implementations for `import_elements`. We need the specialization as long as we don't replace `TBB` with `Kokkos` which I would prefer to fo somewhere else. All your other comments should be addressed.

masterleinad commented:
It seems `Kokkos::Experimental::abs` didn't exist (but `Kokoks::abs` does now).

Rombur commented on include/deal.II/lac/la_parallel_vector.templates.h:
```diff
@@ -372,108 +375,124 @@ namespace LinearAlgebra
+          Kokkos::realloc(indices_dev, tmp_n_elements);
+          Kokkos::deep_copy(indices_dev,
+                            Kokkos::View<size_type *>(indices.data(),
+                                                      tmp_n_elements));
```
`indices` is already a `View` so why do you create a new one? Shouldn't this be a subview?

masterleinad replied:
Fixed.

Rombur commented on include/deal.II/lac/la_parallel_vector.templates.h:
```diff
@@ -372,108 +375,124 @@ namespace LinearAlgebra
-          AssertCuda(error_code);
+          typename ::dealii::MemorySpace::Default::kokkos_space::execution_space
+            exec;
+          Kokkos::parallel_reduce(
```
You forgot to use a name for all the `parallel_reduce`

masterleinad replied:
Fixed.

Rombur reviewed (commented):
Just minor comments. Do you know if this PR fixes the problem we have with ghosted vector (see the cuda/parallel_vector tests failing [here](https://cdash.dealii.org/viewTest.php?onlyfailed&buildid=958))?

masterleinad commented:
> Just minor comments. Do you know if this PR fixes the problem we have with ghosted vector (see the cuda/parallel_vector tests failing [here](https://cdash.dealii.org/viewTest.php?onlyfailed&buildid=958))?

I haven't run the tests on NVIDIA GPUs but only on Intel GPUs (and CPUs using `SYCL`) and those tests were/are passing there.

Rombur commented:
Maybe a problem when we use the CUDA-aware MPI path then. Is there something similar to CUDA-aware MPI with SYCL?

masterleinad commented:
Yes, there is GPU-aware MPI for Intel GPUs but this pull request doesn't address that. That's what #14571 is for (I have only tested that one with `Serial` and `SYCL`+CPU).
